### PR TITLE
feat: Compact now can split compacted results into multiple non-overlapped files based on config max file size

### LINKED ISSUE #4915: Support multiple streams in StreamSplitExec
# Rationale

Currently [StreamSplitExec](https://github.com/influxdata/influxdb_iox/blob/70337087a83827959a88668d2e51fe225833b246/iox_query/src/exec/split.rs#L168) only supports [2 running streams](https://github.com/influxdata/influxdb_iox/blob/70337087a83827959a88668d2e51fe225833b246/iox_query/src/exec/split.rs#L157). Since we want to split a compacted result into many more non-overlapped streams, this ticket is to make those 2 hard-code streams into an array/list of n streams

### PR DESCRIPTION
**Pull request by NGA-TRAN**
Closes #4591 #4790
This  PR includes
- [x] Teach `split_the_stream` to know how to split data into n streams using n-1 input PhysicalExprs
- [x] Make `StreamSplitNode` support a list of expressions
- [x] Teach `compute_split_time` to split a time range into many contiguous ranges
- [x] split compacted result into multiple non-overlapped files based on the config comapction_max_size_bytes

And others
- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: compactor/src/compact.rs [hunk 1]
```diff
@@ -1319,14 +1410,23 @@ mod tests {
             .create_partition("part")
             .await;
         let time = Arc::new(SystemProvider::new());
+        let split_percentage = 90;
```
Comment (NGA-TRAN) [inline compactor/src/compact.rs:1413]: The `split_percentage` is no longer needed in the new split rule. A lot of this left over due to previous setup. I will have a PR to clean this up.

File: compactor/src/compact.rs [hunk 2]
```diff
@@ -1418,20 +1521,24 @@ mod tests {
...
+                partition.partition.id,
+                compact_and_upgrade,
+                compactor.config.compaction_max_size_bytes(),
+            )
             .await
             .unwrap();
 
-        // Should have 3 non-soft-deleted files: pf1 not compacted and stay, and 2 newly created after compacting pf2, pf3, pf4
+        // Should have 2 non-soft-deleted files: pf1 not compacted and stay, and 1 newly created
+        // after compacting pf2, pf3, pf4 all very small into one file
```
Comment (NGA-TRAN) [inline compactor/src/compact.rs:1533]: This whole test is very complicated with all factor considered. Due to new split rule and I want to keep pf2, pf3, pf4 small to tests other factors, the compacted result is now in one file. All the changes below of this test reflect this change, too

**Comment by NGA-TRAN** [comment]
Sorry the PR is a bit long. I hope the bullets for each sub-task help. They are all linked together so I thought instead of faking the function parameters and have 4 small PRs, I put them into one.

File: iox_query/src/frontend/reorg.rs [hunk 3]
```diff
@@ -346,4 +375,133 @@ mod test {
...
+        let split_plan = ReorgPlanner::new()
+            .split_plan(schema, chunks, sort_key, vec![]) // reason of panic: empty split_times
+            .expect("created compact plan");
+
+        let executor = Executor::new(1);
+        let _physical_plan = executor
+            .new_context(ExecutorType::Reorg)
+            .create_physical_plan(&split_plan)
+            .await
+            .unwrap();
```
Comment (alamb) [inline iox_query/src/frontend/reorg.rs:479]:
```suggestion
```

Comment (alamb) [inline iox_query/src/frontend/reorg.rs:479]:
```suggestion
```

File: iox_query/src/frontend/reorg.rs [hunk 4]
```diff
@@ -346,4 +375,133 @@ mod test {
...
+            .split_plan(schema, chunks, sort_key, vec![1000, 500]) // reason of panic: split_times not in ascending order
+            .expect("created compact plan");
+
+        let executor = Executor::new(1);
+        let _physical_plan = executor
+            .new_context(ExecutorType::Reorg)
+            .create_physical_plan(&split_plan)
+            .await
+            .unwrap();
+    }
```
Comment (alamb) [inline iox_query/src/frontend/reorg.rs:506]:
```suggestion
    }
```

File: compactor/src/compact.rs [hunk 5]
```diff
@@ -1740,30 +1873,40 @@ mod tests {
             .await;
 
         // Data: Should have 4 rows left
-        // first set contains least recent 3 rows
+        // first set contains least recent 2 rows
         assert_batches_sorted_eq!(
             &[
```
Comment (alamb) [inline compactor/src/compact.rs:1878]: nice

**Review by alamb — APPROVED** [review]
Looks good to me -- I read the `iox_query` changes quite carefully and they looked good to me. I didn't follow all the compact.rs test changes, but what i did looked good

File: iox_query/src/exec/split.rs [hunk 6]
```diff
@@ -352,113 +344,108 @@ impl StreamSplitExec {
...
-                    filter_record_batch(&batch, &BooleanArray::from(mapped_indicies))
-                } else {
-                    filter_record_batch(&batch, not_true_indices)
-                }?;
-                timer.done();
-
-                (true_batch, not_true_batch)
-            }
-            // handle when the user provided a constant predicate
-            (
```
[the remark this answers is not in the document]
Comment (NGA-TRAN) [reply iox_query/src/exec/split.rs:428]: There are tests for this case `test_constant_split` and `test_constant_split_multi_exprs`

**Pull request merged**